package/tests/test_data_access/test_managers.py

```diff
@@ -375,6 +471,348 @@ def test_add_pipelines(
             ]
         )
 
+    def assert_expected_modular_pipeline_values_for_edge_cases(
+        self,
+        expected_modular_pipeline_tree_obj,
+        modular_pipeline_node_id,
+        data_access_manager,
+        modular_pipeline_tree_values,
+        expected_key,
+    ):
+        """This asserts an `expected_key` value present in modular_pipeline_tree
+        that is constructed in the edge cases with the expected_modular_pipeline_tree"""
+        assert sorted(
+            list(
+                expected_modular_pipeline_tree_obj[modular_pipeline_node_id][
+                    expected_key
+                ]
+            )
+        ) == sorted(
+            list(
+                data_access_manager.nodes.get_node_by_id(node_id).name
+                for node_id in modular_pipeline_tree_values
+            )
+        )
+
+    def test_task_nodes_for_modular_pipeline_edge_cases(
+        self,
+        data_access_manager: DataAccessManager,
+        edge_case_example_pipelines: Dict[str, Pipeline],
+        example_catalog: DataCatalog,
+    ):
+        """
+        This tests the task nodes that are created when
+        edge_case_example_pipelines are added to data access manager
+
+        Testing the use cases mentioned in
+        https://github.com/kedro-org/kedro-viz/pull/1651
+        https://github.com/kedro-org/kedro-viz/issues/1814
+
+        """
+        data_access_manager.add_catalog(example_catalog, edge_case_example_pipelines)
+        data_access_manager.add_pipelines(edge_case_example_pipelines)
+
+        # testing task nodes in GraphNodesRepository
+        all_node_names = set()
+        for registered_pipeline_id, _pipeline in edge_case_example_pipelines.items():
+            # This is needed to add modular pipeline nodes to the GraphNodesRepository
+            data_access_manager.create_modular_pipelines_tree_for_registered_pipeline(
+                registered_pipeline_id
+            )
+            all_node_names |= {node.name for node in _pipeline.nodes}
+
+        assert sorted(
+            {
+                f"{n.namespace}.{n.name}" if n.namespace else n.name
+                for n in data_access_manager.nodes.as_list()
+                if n.type == GraphNodeType.TASK
+            }
+        ) == sorted(list(all_node_names))
```

Comment:
I find this test quite hard to follow because both sides of the assert are generated in the test. If you look at other tests in this file e.g. https://github.com/kedro-org/kedro-viz/blob/main/package/tests/test_data_access/test_managers.py#L346-L354 you see that one side of the assert is generated whereas the other is very explicit, which makes it easier to reason about IMO.